Dockerfile

```diff
@@ -1,89 +1,43 @@
-mv /usr/local/clang+llvm-3.8.1-x86_64-linux-gnu-ubuntu-16.04 /usr/local/clang+llvm && \
-rm clang+llvm.tar.xz && \
-rm -fr /usr/local/clang+llvm/include/llvm-c && \
-rm -fr /usr/local/clang+llvm/include/clang-c && \
-rm -fr /usr/local/clang+llvm/include/c++ && \
-rm -fr /usr/local/clang+llvm/share && \
-ls -d /usr/local/clang+llvm/lib/* | grep -vE clang$ | xargs rm -r && \
-ls -d /usr/local/clang+llvm/bin/* | grep -vE "clang$|clang-3.8$|llc$" | xargs rm -r && \
-# clang-3.8.1-end
-
-# iproute2-begin
-cd /tmp && \
-git clone -b v4.10.0 git://git.kernel.org/pub/scm/linux/kernel/git/shemminger/iproute2.git && \
-cd /tmp/iproute2 && \
-./configure && \
-make -j `getconf _NPROCESSORS_ONLN` && \
-make install && \
-# iproute2-end
-
-# bpf-map-begin
-curl -SsL https://github.com/cilium/bpf-map/releases/download/v1.0/bpf-map -o bpf-map && \
-chmod +x bpf-map && \
-mv bpf-map /usr/bin && \
-# bpf-map-end
-
-# cni-begin
-#Include the loopback binary in the image
-mkdir -p tmp && cd tmp && \
-curl -sS -L https://storage.googleapis.com/kubernetes-release/network-plugins/cni-0799f5732f2a11b329d9e3d51b9c8f2e3759f2ff.tar.gz -o cni.tar.gz && \
-tar -xvf cni.tar.gz && \
-mkdir /cni && \
-cp bin/loopback /cni && \
-cd .. && \
-rm -r tmp && \
-# cni-end
-
-cd /tmp && \
-curl -Sslk -o go.linux-amd64.tar.gz \
-https://storage.googleapis.com/golang/go1.9.linux-amd64.tar.gz && \
-tar -C /usr/local -xzf go.linux-amd64.tar.gz && \
-cd /tmp/cilium-net-build/src/github.com/cilium/cilium && \
-export GOROOT=/usr/local/go && \
-export GOPATH=/tmp/cilium-net-build && \
-export PATH="$GOROOT/bin:/usr/local/clang+llvm/bin:$GOPATH/bin:$PATH" && \
-go get -u github.com/jteeuwen/go-bindata/... && go get -u github.com/google/gops && \
-mv /tmp/cilium-net-build/bin/gops /usr/local/bin/ && \
-cd /tmp/cilium-net-build/src/github.com/cilium/cilium && \
-export GOROOT=/usr/local/go && \
-export GOPATH=/tmp/cilium-net-build && \
-export PATH="$GOROOT/bin:/usr/local/clang+llvm/bin:$GOPATH/bin:$PATH" && \
-make clean-container build && \
-make PKG_BUILD=1 install && \
-
-# bash-completion-begin
-mkdir -p /root && \
-echo ". /etc/profile.d/bash_completion.sh" >> /root/.bashrc && \
-cilium completion bash >> /root/.bashrc && \
-# bash-completion-end
-
-groupadd -f cilium && \
-
-apt-get purge --auto-remove -y gcc make bison flex git curl xz-utils ca-certificates && \
-apt-get clean && \
-rm -fr /root /var/lib/apt/lists/* /tmp/* /var/tmp/* /usr/local/go
+#
+# Cilium incremental build. Should be fast given builder-deps is up-to-date!
+#
+# cilium-builder tag is the date on which the compatible build image
+# was pushed.  If a new version of the build image is needed, it needs
+# to be tagged with a new date and this file must be changed
+# accordingly.  Keeping the old images available will allow older
+# versions to be built while allowing the new versions to make changes
+# that are not backwards compatible.
+#
+FROM cilium/cilium-builder:2017-12-14 as builder
```

Comment:
I'd suggest the tag also include a revision for that day. This allows us to push hotfixes to images more than once a day, and then avoid ambiguity over which actual image is being run on machines during these bugfix pushes. An alternative would be to have hour:minute on there too but that has a similar problem on a minute time scale.

Reply:
Builder and runtime base images are not very likely to be re-created very often. Especially for the builder image I do not see a scenario where hotfixes would be needed, as bazel build cache will just become less useful it becomes out-of-date.

Generally, using the git sha as the tag would be most accurate, but even that does not cover the fact that some runtime and build tool dependencies are not explicitly managed, but we install the latest version available via apt.

So I though using a date would strike a good balance here, implicitly conveying a notion of freshness. Nothing prevents adding retagging the images manually after a build with a more detailed tag either.

That said, the above is all there is to the choice of the automatically generated tag, open to more discussion and viewpoints as always :-)

Reply:
This will totally work as-is. I was projecting some mostly forgotten issue I once had (all I do remember is having to retrofit the revision onto the date tag). In this case, especially, it's pretty easy to fix it with the manual tagging, as you said. +1